### PR TITLE
crimson/osd: add crush_location_hook feature to crimson

### PR DESCRIPTION
**Pull request by rosinL**
crimson/osd: add crush_location_hook feature to crimson

Signed-off-by: luo rixin <luorixin@huawei.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by rosinL** [comment]
@athanatos , Hi Sam, I am porting crush_location_hook feature to crimson, please help to review.

**Review by Matan-B — COMMENTED** [review]
I didn't go over in detail yet. ~However, please see make check results for `TestCephAdm.test_mon_crush_location`.~
Edit: This test seem to fail unrelated to the changes here, you can ignore this comment.

**Comment by rosinL** [comment]
jenkins retest this please

File: src/crush/CrushLocation.cc [hunk 1]
```diff
@@ -43,14 +45,120 @@ int CrushLocation::_parse(const std::string& s)
...
+  }
+}
+
+using namespace crimson::common;
+seastar::future<> CrushLocation::update_from_hook()
+{
+  if (local_conf().get_val<std::string>("crush_location_hook").length() == 0)
+    return seastar::now();
+
+  return seastar::file_exists(local_conf().get_val<std::string>("crush_location_hook")).then([this] (bool result) {
```
Comment (Matan-B) [inline src/crush/CrushLocation.cc:60]: Formatting, please split this line (and the other long ones).
https://google.github.io/styleguide/cppguide.html#Line_Length

File: src/crush/CrushLocation.cc [hunk 2]
```diff
@@ -43,14 +45,120 @@ int CrushLocation::_parse(const std::string& s)
...
+  if (r < 0)
+    strcpy(hostname, "unknown_host");
+  // use short hostname
+  for (unsigned i=0; hostname[i]; ++i) {
+    if (hostname[i] == '.') {
+      hostname[i] = '\0';
+      break;
+    }
+  }
+  std::lock_guard l(lock);
```
Comment (Matan-B) [inline src/crush/CrushLocation.cc:149]: Crimson uses a dummy_mutex. We can skip the lock_guard.

**Comment by rosinL** [comment]
jenkins retest this please

File: src/crush/CrushLocation.cc [hunk 3]
```diff
@@ -43,14 +45,120 @@ int CrushLocation::_parse(const std::string& s)
   return 0;
 }
 
+#if defined(WITH_SEASTAR) && !defined(WITH_ALIEN)
```
[the remark this answers is not in the document]
Comment (rosinL) [reply src/crush/CrushLocation.cc]: done.

File: src/crush/CrushLocation.cc (hunk 1 above)
Comment (rosinL) [reply src/crush/CrushLocation.cc]: done

File: src/crush/CrushLocation.cc (hunk 2 above)
Comment (rosinL) [reply src/crush/CrushLocation.cc]: done

**Comment by rosinL** [comment]
jenkins retest this please

**Review by athanatos — APPROVED** [review]
LGTM, needs to go through a normal qa run since it touches crush/CrushLocation.h/cc

**Comment by ljflores** [comment]
Rados suite review: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcJ04nAx3y1756-wip-yuri11-testing-2023-05-19-0836

**Pull request merged**